src/Kestrel.Core/Internal/HttpConnection.cs

```diff
@@ -143,15 +144,30 @@ public Task StartRequestProcessing<TContext>(IHttpApplication<TContext> applicat
                     adaptedPipelineTask = adaptedPipeline.RunAsync(stream);
                 }
 
-                if (_http1Connection.ConnectionFeatures.Get<ITlsApplicationProtocolFeature>()?.ApplicationProtocol == "h2" &&
-                    Interlocked.CompareExchange(ref _http2ConnectionState, Http2ConnectionStarted, Http2ConnectionNotStarted) == Http2ConnectionNotStarted)
+                var hasTls = _context.ConnectionFeatures.Get<ITlsConnectionFeature>() != null;
+                var applicationProtocol = _context.ConnectionFeatures.Get<ITlsApplicationProtocolFeature>()?.ApplicationProtocol;
+                var http1Enabled = (_context.Protocols & EndPointProtocols.Http1) == EndPointProtocols.Http1;
+                var http2Enabled = (_context.Protocols & EndPointProtocols.Http2) == EndPointProtocols.Http2;
+
+                if (!hasTls && http1Enabled && http2Enabled)
+                {
+                    throw new InvalidOperationException("HTTP/1.x and HTTP/2 cannot both be served on the same cleartext endpoint.");
+                }
+
+                var useHttp2 = http2Enabled && (!hasTls || applicationProtocol == "h2");
+
+                if (useHttp2 && Interlocked.CompareExchange(ref _http2ConnectionState, Http2ConnectionStarted, Http2ConnectionNotStarted) == Http2ConnectionNotStarted)
                 {
                     await _http2Connection.ProcessAsync(httpApplication);
                 }
-                else
+                else if (http1Enabled && (applicationProtocol == null || applicationProtocol == "http/1.1"))
                 {
                     await _http1Connection.ProcessRequestsAsync();
                 }
+                else
+                {
+                    throw new InvalidOperationException("Unable to select a protocol for the endpoint.");
```

Comment:
I generally dislike "Unable to do thing" messages that don't say why. If a few more `if` clauses are needed to figure out a more specific error, I say we do that work so consumers don't have to.